### DISCUSSION TITLE
Interactive Drawing Tools Build Complete - Plugin Conversion Needed

### DISCUSSION THREAD

**Discussion by difurious** [tradingview/lightweight-charts#1466] [General]
I have a custom lightweight-charts build based off of v3.8 that lets you create interactive drawing tools on the chart.  I call it the Line Tools build, project and video located [here.](https://github.com/difurious/lightweight-charts-line-tools#video)  

Tools that can be created are FibRetracement, ParallelChannel, HorizontalLine, VerticalLine, Highlighter, CrossLine, TrendLine, Rectangle, Triangle, Brush, Path, Text, Ray, Arrow, ExtendedLine, HorizontalRay, Circle, Callout, PriceRange

It works great and I need help from someone to update it to V4.x.x and if possible to a plugin.  If you have the skill, this would make an awesome open source plugin for the community. Let me know if you want to take on this challenge.

**Comment by difurious** [comment]
just added Circle, Callout, PriceRange .  Let me know if you want to help convert into Plugin.

**Reply by sydrock1204** [reply]
> just added Circle, Callout, PriceRange . Let me know if you want to help convert into Plugin.

You are great. Please Know each other. @mihailflorin is my telegram id. 
+381628645113 is my whatsapp
live:.cid.e13d554e3ea65c2c is my skype.

**Comment by tcnoff** [comment]
This is fantastic. Would love to see this implemented!

**Comment by brlocky** [comment]
Hi all !
I'm also looking for a toolset to use with lightweight charts, we developed some plugins like the Fibonacci retracement and projection tools, they work but need polish.
Anyways just to share some plugin examples


Would be interested also in Object Tree to manage the objects

https://github.com/brlocky/Elliott-UI/tree/main/src/components/Chart/plugins/fibonacci-drawing-tool

https://github.com/brlocky/Elliott-UI/tree/main/src/components/Chart/plugins/fibonacci-projection-tool

**Comment by TiS-QSP** [comment]
Hi @difurious, great work on your previous implementation BTW, thanks a lot. What is the status of this? Have you already created any plugins?

**Reply by difurious** [reply]
I have not converted it to plugins.  Not sure if I will get to that.  I am using it daily and works great.  Downside is that it is on 3.8 of lightweight.  Was hoping someone would use the tools as inspiration on what can be done and make plugins out of each tool.  I imagined a plugin for core functionality, and individual plugins for each tool so you can mix and match.  So anyone wanting to put in the final push to make this more future proof, let me know.

**Comment by KingOfFibonacci** [comment]
Just finding this now after wrestling with implementing my own. Thank you sir! Will let you know how implementing this goes into the current version...

**Reply by difurious** [reply]
im currently making it into a V5 plugin(s).  Im maybe a month + out to having something. It going to be hopefully something people will help build new additional tools off of it.

**Reply by crypt0inf0** [reply]
Waiting for V5

**Reply by aaahariri** [reply]
@difurious Also working on a new project and very interested in v5. Happy to donate a bit too as a token of appreciation. Is there a way for us to subscribe to updates (Not that used to Github).

**Comment by KingOfFibonacci** [comment]
How we looking boys?
<img width="1908" height="988" alt="image" src="https://github.com/user-attachments/assets/10e650a4-0665-458c-a856-25ea33d6cdce" />

**Reply by bizzaretrading** [reply]
neat.

**Comment by crypt0inf0** [comment]
check out https://github.com/crypt0inf0/lightweight-charts-ui & demo available at https://crypt0inf0.github.io/lightweight-charts-demo/

All the drawing tools plugin available at https://github.com/crypt0inf0/lightweight-charts/tree/master/plugin-examples/src/plugins/line-tools
try ,
```
cd plugin-examples/src/plugins/line-tools

npm i

npm run build 

npm run dev
```
demo available at http://localhost:5173/src/plugins/line-tools/example/index.html

**Reply by difurious** [reply]
cool demo, your demo page seems to be ui only and no functional tools?.  Is your line tools based on my v3.8?  Im probably a week+ or so out on releasing my v5 conversion  for the tools in my v3.8 build.  Mine is lacking a ui to create the tools. I made a test frontend app to test all the tools functionality, but your gui would be an excellent addition to what im cooking.  The core plugin is the nuts and bolts and you can add ala cart the individual line tools.  I will let you know when its released

**Reply by crypt0inf0** [reply]
Yes i created my tools based on your tools and i use UI directly for smooth experience. 

Added credit https://github.com/crypt0inf0/lightweight-charts-ui/blob/main/README.md#credit

**Reply by bizzaretrading** [reply]
this is really good. I had almost given up on light weight charts becuase the drawings won't extend into any future date (beyond what data is available on the chart). I was going to refactor my almost near ready product to use klinecharts library and today you drop this. I cannot thank you guys enough

**Reply by difurious** [reply]
> Yes i created my tools based on your tools and i use UI directly for smooth experience.
> 
> Added credit https://github.com/crypt0inf0/lightweight-charts-ui/blob/main/README.md#credit

awesome, there is a chance my build will be done either in days, or delayed by a week + because of the holidays.  But i imagine your ui being an addition to the "core" so it makes things easier to create the tools.  eventually a ui to modify the options of the tools when you double click.  my build is meant to be a drop in replacement to my existing trading tool as much as possible, so that is why i need to make the tool as similar as possible to v3.8 build.  I will let you know when it is ready so you can see if it possible to make a "plugin" for my "core" to have it be like the ui for the tools.

**Reply by sasuke4ever** [reply]
@crypt0inf0 Your project is impressive, and the custom tools you've built are excellent. Would you consider creating a video tutorial on YouTube explaining your development process? A guide on how to build such tools would be incredibly helpful for me and other programmers looking to create our own.

**Comment by vishnuc** [comment]
why cant I drag drawings

**Reply by crypt0inf0** [reply]
which plugin are you using?

**Reply by vishnuc** [reply]
trendline , fibonacci etc

**Comment by bizzaretrading** [comment]
Feature Request: exportTools() / importTools() methods for LineToolManager
Summary
Add methods to serialize and deserialize all drawing tools managed by LineToolManager, enabling persistence across page reloads and symbol changes in multi-chart applications.

Use Case
I'm building a trading dashboard with multiple charts where users frequently switch between symbols. When switching symbols, I need to:

Save drawings for the current symbol
Restore drawings when the user switches back


Currently, I work around this by caching the actual tool primitives in memory and detaching/re-attaching them:
````
// On symbol change - cache tools
const tools = [...manager._tools];
symbolCache.set(prevSymbol, tools);
tools.forEach(t => series.detachPrimitive(t));
manager._tools.length = 0;

// On symbol return - restore tools
const cached = symbolCache.get(newSymbol);
cached.forEach(t => {
  series.attachPrimitive(t);
  manager._tools.push(t);
});
```
Problem: This only works within a single session. On page refresh, the in-memory cache is lost. Users lose all their drawings.

`// Export all tools as serializable JSON
const data = manager.exportTools();
// → [{ toolType: 'TrendLine', p1: {logical, price}, p2: {logical, price}, options: {...} }, ...]

localStorage.setItem('drawings', JSON.stringify(data));

// Later, restore tools
const saved = JSON.parse(localStorage.getItem('drawings'));
manager.importTools(saved);`


Why this matters
Drawing persistence is essential for trading applications. Users spend significant time annotating charts with trendlines, Fibonacci retracements, and support/resistance levels. Losing these on refresh creates a poor user experience.

Current Workarounds Attempted
In-memory cache - Works for session, lost on refresh
Accessing manager._tools - Can read primitives but can't recreate them from serialized data since tool constructors (TrendLine, Rectangle, etc.) aren't exported
Environment
lightweight-charts-ui version: [your version]
lightweight-charts version: [your version]
Browser: Chrome/Firefox/etc.
Thank you for this great library! The drawing tools, selection handles, and properties dialogs work beautifully. This feature would complete the persistence story. Happy to help test or contribute if you point me in the right direction.

**Reply by bizzaretrading** [reply]
if you could point me In right direction I would write it myself @difurious

**Reply by difurious** [reply]
crypt0inf0's build is not what im building.  In mine, I do have an export, import all tools methods.  I am getting close with my build to be done so you will have to wait a bit.

**Reply by bizzaretrading** [reply]
I will happily wait

**Reply by crypt0inf0** [reply]
I also wait and update my code to support export and import methods

**Reply by tedydonel** [reply]
@difurious @crypt0inf0 etc........ I don't know how this thing actually works. this is my first time ever in a discussion, but that's that important. I just modified alot from your build [https://github.com/crypt0inf0/lightweight-charts-ui](https://github.com/crypt0inf0/lightweight-charts-ui) 
I modified the positions tools 80% close to that of tradingview which works well with your replay option, and i also added pairs from trading view. 
<img width="1305" height="632" alt="Screenshot 2025-12-17 231919" src="https://github.com/user-attachments/assets/bcc52bde-625c-4877-a463-d75e2e0c7413" />
<img width="1352" height="602" alt="image" src="https://github.com/user-attachments/assets/efd5d52c-2c92-4ea0-b895-27cf0ff52dd5" />

**Reply by tedydonel** [reply]
Ohh, I didn't commit yet, tell me what you think?

**Reply by crypt0inf0** [reply]
Fork 
https://github.com/crypt0inf0/lightweight-charts-ui and commit your changes so that we can see what improvements you made.

**Reply by tedydonel** [reply]
Yeah, I wanna make some more modifications before committing

**Reply by tedydonel** [reply]
@crypt0inf0 I forked it [https://github.com/tedydonel/lightview](https://github.com/tedydonel/lightview)
<img width="998" height="577" alt="OANDA_XAUUSD" src="https://github.com/user-attachments/assets/7d404a13-2bf5-4a86-ae92-2e4cc1e40e0c" />

**Reply by crypt0inf0** [reply]
i will check it out

**Comment by difurious** [comment]
I'm out for the holidays for a week. I have the core and all tools created and tested   only thing that could be better is the mouse pointer changing to different pointers conditionally.  Just cosmetic.  But my goal is a drop in replacement for the 3.8 build. I have a test app that tests all tools and option in a visual way so it gives examples on how everything works.  The remaining thing is documentation and readme and upload to GitHub.  So when I get back, hopefully less than a week to have V1 done.  The core is set up to be like an api for the tools to use.  So if someone wants to make their own tools it takes care of a lot of things.  I don't imagine taking requests for new tools.  But for you to use the core and make your own and release the tools for all to use.

**Comment by tedydonel** [comment]
What would it take to build a complete industry standard charting library like trading view from using the it's lightweight library?

**Comment by difurious** [comment]
my build of the line tools build converted into lightweight-charts V5+ plugins is done. [ Here is the link to the core plugin](https://github.com/difurious/lightweight-charts-line-tools-core) . The core is the brains and each line tool registers with the core. Make sure to read the whole readme.  Here is a list of all the line tools that I have created. 

Use the core to make your own line tool,  but make sure to share it publicly with the community. 

- TrendLine
- Ray
- Arrow
- ExtendedLine
- HorizontalLine
- HorizontalRay
- VerticalLine
- CrossLine
- Callout
- Brush
- Highlighter
- Rectangle
- Circle
- Triangle
- Path
- ParallelChannel
- FibRetracement
- PriceRange
- LongShortPosition
- Text
- MarketDepth

**Reply by crypt0inf0** [reply]
I bundled pulgin and testing together for quick testing and development on https://github.com/crypt0inf0/lightweight-charts-plugins

**Reply by difurious** [reply]
the test app in the packages.json already has all the plugins as dependencies, so the user just needs to do "npm install".  It will fetch and build all plugins due to each plugin building automatically.  

now that the 21 different line tools is out, would it be possible for your uI to be like a plugin to potentially read from the core what plugins are "registered" with the core and then the UI would display the tool bar so there is an easy ui to click and create said tool.

Seems like some method in the core would need to be created to read what tools are roistered to get the "names" of the registered tools, then your ui reads that and makes the UI, so your UI is dynamic and based off of only what is loaded.  Thoughts on this?

**Reply by crypt0inf0** [reply]
I already use your plugins with my UI everything works fine but floating toolbar where we change color and so on is not working as expected. I moved it to frontend to strictly keep plugins separated. I'm working on it to make it smooth.

**Reply by difurious** [reply]
you did have your own converted versions of the 3.8 plugins, but now you are using my version of the conversion?

yeah the changing options is hard.  my tool i have I have a custom bit of frontend to handle that.  it reads the info coming from the subscribeLineToolsDoubleClick and since all the info about the tool is passed from the backend, I can read it to know what things are possible for the frontend to generate to make the UI for the options modification.  Its not updating the tools option on each change, but you submit the form and then it validates the form fields to make sure they are valid, then updates the tool on submit.  

The main things that would be useful to people are the menu bar to click the tools icon to create an interactive tool, and the options ui to change the double clicked tool.  Your ui build had a lot of other lightweight functionality unrelated to line tools, but the 2 i mentioned would help make line tools more complete because it is now user friendly without having to program anything.

**Reply by crypt0inf0** [reply]
It is extremely easy to switch from my plugins to your modular plugins. The problem is linking drawing with its edit tool navbar. It works but when I move and try to lock or unlock it. The edit toolbar gets disconnected from the drawing and I see nothing happens. Without moving the drawing I can change colour, width, lock/unlock, delete it. But when I move it nothing works. It's a frontend issue.

**Reply by crypt0inf0** [reply]
@difurious I tried building a floating navbar on the frontend for your plugins but it is not smooth unable to bridge between drawings and the navbar in the frontend. I'm going to stick with my combined plugin.

**Reply by jusacc** [reply]
Great job! May I ask you to add one tool - time circles (it looks like semicircles). Thx a lot

**Reply by sasuke4ever** [reply]
This is absolutely perfect—exactly what I’ve been looking for over the past two months! I can’t thank you enough.

**Reply by sasuke4ever** [reply]
@difurious  Could you also create a tutorial on how to design custom drawing tools?

**Reply by difurious** [reply]
> @difurious Could you also create a tutorial on how to design custom drawing tools?

The tutorial is in the source code. If you build it build docs for the api.  you need to look at the source to find the notes on how things work.  Im not doing more effort tutorial wise that what is already out there for you.  feed the project to the AI and it will be able to answer any questions you want.

**Reply by difurious** [reply]
> Great job! May I ask you to add one tool - time circles (it looks like semicircles). Thx a lot

Im not taking requests for tools.  but the circle plugin is 99% what you need, just look at the draw method and instead of it drawing a full circle have it draw half a circle. And rename the tool to something else, then publish on git for all to use.

**Reply by jusacc** [reply]
> Im not taking requests for tools. but the circle plugin is 99% what you need, just look at the draw method and instead of it drawing a full circle have it draw half a circle. And rename the tool to something else, then publish on git for all to use.

There is an important detail that the selected cycle period has an infinite continuation both forward and backward.. so I'm afraid it's unlikely that the drawing tool will work this way. Unfortunately, I'm not a developer in any way, so I won't be able to create this tool on my own. It remains to be hoped that someone from the community will do this. In any case.. Thank you for your work.

**Comment by SilentHG** [comment]
I am using the CDN js package, I am backend guy, any support for guys like me ? how do i integrate into my charts ? all i am seeing are .ts files, my stack isn't node. backend is python, frontend is good old Html+css+js

**Reply by difurious** [reply]
Im assuming you are not using the Old v3.8 build , but the new plugin version, this one https://github.com/tradingview/lightweight-charts/discussions/1466#discussioncomment-15314226 .  you need to build the source, it will make the dist folder, and will make .js files you can import into your frontend because its just javascript.

**Comment by uamopho** [comment]
The only bug I got using diffurious tool is on mobile, i need help
